tensorflow/core/kernels/data/shuffle_dataset_op.cc

```diff
@@ -12,718 +12,731 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
+#include "tensorflow/core/kernels/data/shuffle_dataset_op.h"
 
 #include <deque>
 #include <vector>
 
-#include "tensorflow/core/framework/dataset.h"
 #include "tensorflow/core/framework/partial_tensor_shape.h"
 #include "tensorflow/core/framework/resource_mgr.h"
 #include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/kernels/data/name_utils.h"
 #include "tensorflow/core/lib/random/philox_random.h"
 #include "tensorflow/core/lib/random/random.h"
 #include "tensorflow/core/lib/random/random_distributions.h"
 
 namespace tensorflow {
 namespace data {
-namespace {
-
-const int64 kLogIntervalMicros = 10 * 1000000;  // 10 seconds.
-
-const int64 kMaxEpochsInBuffer = 3;
 
 // See documentation in ../../ops/dataset_ops.cc for a high-level
 // description of the following op.
 
-class ShuffleDatasetOpBase : public UnaryDatasetOpKernel {
+/* static */ constexpr const char* const ShuffleDatasetOpBase::kInputDataset;
+/* static */ constexpr const char* const ShuffleDatasetOpBase::kBufferSize;
+/* static */ constexpr const char* const ShuffleDatasetOpBase::kSeed;
+/* static */ constexpr const char* const ShuffleDatasetOpBase::kSeed2;
+/* static */ constexpr const char* const ShuffleDatasetOpBase::kOutputTypes;
+/* static */ constexpr const char* const ShuffleDatasetOpBase::kOutputShapes;
+
+/* static */ constexpr const char* const ShuffleDatasetOp::kDatasetType;
+/* static */ constexpr const char* const
+    ShuffleDatasetOp::kReshuffleEachIteration;
+
+/* static */ constexpr const char* const
+    ShuffleAndRepeatDatasetOp::kDatasetType;
+/* static */ constexpr const char* const ShuffleAndRepeatDatasetOp::kCount;
+
+const int64 kLogIntervalMicros = 10 * 1000000;  // 10 seconds.
+const int64 kMaxEpochsInBuffer = 3;
+
+constexpr char kNumRandomSamples[] = "num_random_samples";
+constexpr char kEndOfInputSequence[] = "end_of_input_sequence";
+constexpr char kEpoch[] = "epoch";
+constexpr char kNumElements[] = "num_elements";
+constexpr char kSlicesSize[] = "slices_size";
+constexpr char kSlicesStart[] = "slices_start_";
```

Comment:
optional: I would prefer to avoid using trailing or leading `_`, instead create a utility method that takes the variable (e.g. slice index or buffer index and generates the right string)

Reply:
Good idea! Add a util function `string StrJoin(const string& separator, const Args&... args)` to generate the right string.